### PR TITLE
References feature

### LINKED ISSUE #4574: Implement Find References language feature
**Description:**
Add the language feature [Find References](https://microsoft.github.io/language-server-protocol/specification#textDocument_references) for ballerina language server.

### PR DESCRIPTION
**Pull request by NipunaMarcus**
## Goals
This will introduce the initial implementation of "Find all references" [1] [2] language feature to the ballerina language server. For now this supports following constructs of the ballerina language.

- Functions 
- Connectors initial support
- Actions initial support
- Structs initial support
- Enum initial support
- Global and local variable initial support 

There are few scenarios which are still not supporting, like search inside if condition, while condition, endpoint creation. Support for these will be added soon with future releases. 

*Note: There is an issue with position calculation of nodes as some nodes (i.e. Enum) contains -1 as column values. So highlighting of some references will be wrong until fixing this issue.* 

### Functions
ezgif com-video-to-gif 15

### Connectors
ezgif com-video-to-gif 21

### Action
ezgif com-video-to-gif 22

### Structs
ezgif com-video-to-gif 16

### Enums
ezgif com-video-to-gif 20

### Local Variable
ezgif com-video-to-gif 17

### Global Variable
ezgif com-video-to-gif 18

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/BallerinaTextDocumentService.java [hunk 1]
```diff
@@ -209,7 +211,28 @@ public BallerinaTextDocumentService(BallerinaLanguageServer ballerinaLanguageSer
 
     @Override
     public CompletableFuture<List<? extends Location>> references(ReferenceParams params) {
-        return CompletableFuture.supplyAsync(() -> null);
+        return CompletableFuture.supplyAsync(() -> {
+            ReferenceParams ref = params;
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/BallerinaTextDocumentService.java]: Shall we remove the redundant local variable?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/position/PositionTreeVisitor.java [hunk 2]
```diff
@@ -151,17 +116,24 @@ public void visit(BLangImportPackage importPkgNode) {
...
-
     public void visit(BLangFunction funcNode) {
         // Check for native functions
         BSymbol funcSymbol = funcNode.symbol;
         if (Symbols.isNative(funcSymbol)) {
             return;
         }
 
+        if (funcNode.getPosition().sLine == this.position.getLine() && funcNode.getPosition().sCol == 1) {
+            this.context.put(NodeContextKeys.NODE_KEY, funcNode);
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/position/PositionTreeVisitor.java]: Let's try to have the following context populating code segment to a util within the tree visitor (Just a suggestion :) ).

Comment (NipunaMarcus) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/position/PositionTreeVisitor.java]: it'll be good if we move this to a util. Let's do a full fetch design and code review on ballerina language server and try to identify what can be reuse and how they can be placed as now we have almost all the important language features .

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/BallerinaTextDocumentService.java (hunk 1 above)
Comment (NipunaMarcus) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/BallerinaTextDocumentService.java]: will do ... thank you

**Pull request merged**